cynic-codegen/src/fragment_derive/fragment_impl.rs

```diff
@@ -204,31 +206,47 @@ impl quote::ToTokens for FieldSelection<'_> {
             }
         });
 
+        let selection_mode = match (&self.graphql_field_kind, self.flatten, self.recurse_limit) {
+            (FieldKind::Enum | FieldKind::Scalar, true, _) => SelectionMode::FlattenLeaf,
+            (FieldKind::Enum | FieldKind::Scalar, false, _) => SelectionMode::Leaf,
+            (_, true, None) => SelectionMode::FlattenComposite,
+            (_, false, None) => SelectionMode::Composite,
+            (_, false, Some(limit)) => SelectionMode::Recurse(limit),
+            _ => panic!("Uncertain how to select for this field."),
+        };
+
+        let aligned_type = match selection_mode {
+            SelectionMode::Composite | SelectionMode::Leaf => {
+                // If we're doing a normal select we need to align types.
+                types::align_output_type(
+                    &types::parsing2::parse_rust_type(field_type),
+                    &self.graphql_field.field_type,
+                )
+                .to_syn()
```

Comment:
This could be much nicer - probably want to make the align_x_type functions take a syn::Type and do the parsing themselves